ra-optimisation/search-tree-rao/src/main/java/com/powsybl/openrao/searchtreerao/castor/algorithm/CastorFullOptimization.java

```diff
@@ -736,13 +729,133 @@ static void applyPreventiveResultsForAutoOrCurativeRangeActions(Network network,
+                }
+            }));
+        });
+        multipleInstantRangeActions.forEach(ra -> {
+            setPointResults.get(preventiveState).put(ra, firstPreventiveResult.getOptimizedSetpoint(ra, preventiveState));
+            contingencyResults.forEach((state, result) -> {
+                if (isRangeActionAvailableInState(ra, state, crac)) {
+                    setPointResults.putIfAbsent(state, new HashMap<>());
+                    setPointResults.get(state).put(ra, result.getOptimizedSetpoint(ra, state));
+                }
+            });
+        });
+        return setPointResults;
+    }
+
+    /**
+     * Checks if raUsageLimits are at risk if we choose to re-optimize a range action.
+     * Returns True if it's at risk, False otherwise.
+     */
+    private static boolean shouldRemoveRaDueToUsageLimits(String operator, RaUsageLimits raUsageLimits, Set<RemedialAction<?>> potentiallyActivatableRas) {
+        long potentiallyActivatableTsos = Integer.MAX_VALUE;
+        long potentiallyActivatableRasForTheTso = Integer.MAX_VALUE;
+        if (operator != null) {
+            potentiallyActivatableRasForTheTso = potentiallyActivatableRas.stream()
+                .filter(ra -> operator.equals(ra.getOperator()))
+                .count();
+            potentiallyActivatableTsos = potentiallyActivatableRas.stream()
+                .map(RemedialAction::getOperator)
+                .filter(Objects::nonNull)
+                .distinct()
+                .count();
+        }
+        int limitingValueForTheTso = Math.min(raUsageLimits.getMaxPstPerTso().getOrDefault(operator, Integer.MAX_VALUE), raUsageLimits.getMaxRaPerTso().getOrDefault(operator, Integer.MAX_VALUE));
+        return raUsageLimits.getMaxRa() < potentiallyActivatableRas.size()
+            || limitingValueForTheTso < potentiallyActivatableRasForTheTso
+            || raUsageLimits.getMaxTso() < potentiallyActivatableTsos;
+    }
+
+    /**
+     * Gathers every range action that should not be considered in the second preventive if those 2 criterion are met :
+     * 1- The range action has the same tap in preventive and in a contingency scenario.
+     * 2- For the given state, the crac has limiting RaUsageLimits.
+     */
+    private static Set<RangeAction<?>> getRangeActionsToRemove(Crac crac, State preventiveState, Map<State, Map<RangeAction<?>, Double>> setPointResults, Map<State, OptimizationResult> contingencyResults) {
+        Set<RangeAction<?>> rangeActionsToRemove = new HashSet<>();
+        setPointResults.forEach((state, spMap) -> spMap.forEach((ra, setPoint) -> {
+            double referenceSetPoint = setPointResults.get(preventiveState).get(ra);
+            if (!state.isPreventive() && setPoint.equals(referenceSetPoint) && crac.getRaUsageLimitsPerInstant().containsKey(state.getInstant())) {
+                Set<RemedialAction<?>> potentiallyActivatableRas = new HashSet<>(crac.getPotentiallyAvailableRangeActions(state));
```

Comment:
you can move this line inside the first forEach to avoid getting it for each ra

Reply:
Yes I realized that and I'm changing this.